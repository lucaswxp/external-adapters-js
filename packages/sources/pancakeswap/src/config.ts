import { Requester, util } from '@chainlink/ea-bootstrap'
import { Config as BaseConfig, ConfigFactory } from '@chainlink/types'
import { ethers } from 'ethers'

export const NAME = 'PANCAKESWAP'

export const ENV_RPC_URL = 'RPC_URL'
export const ENV_ADDRESS_PROVIDER = 'ADDRESS_PROVIDER'
export const ENV_BLOCKCHAIN_NETWORK = 'BLOCKCHAIN_NETWORK'

export const DEFAULT_BLOCKCHAIN_NETWORK = 'bsc'
export const DEFAULT_ENDPOINT = 'crypto'
export const ROUTER_CONTRACT = '0x10ED43C718714eb63d5aA57B78B54704E256024E'

export type Config = BaseConfig & {
  provider: ethers.providers.Provider
}

export const makeConfig: ConfigFactory<Config> = (prefix) => {
  return {
    ...Requester.getDefaultConfig(prefix),
    defaultEndpoint: DEFAULT_ENDPOINT,
    network: util.getEnv(ENV_BLOCKCHAIN_NETWORK, prefix) || DEFAULT_BLOCKCHAIN_NETWORK,
    provider: new ethers.providers.JsonRpcProvider(util.getRequiredEnv(ENV_RPC_URL, prefix)),
  }
}
