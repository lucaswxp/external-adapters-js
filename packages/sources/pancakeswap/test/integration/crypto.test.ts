import { AdapterRequest } from '@chainlink/types'
import request, { SuperTest, Test } from 'supertest'
import process from 'process'
import nock from 'nock'
import http from 'http'
import { server as startServer } from '../../src'
import { mockEthereumResponseSuccess } from './fixtures'
import { AddressInfo } from 'net'

let oldEnv: NodeJS.ProcessEnv

beforeAll(() => {
  oldEnv = JSON.parse(JSON.stringify(process.env))
  process.env.CACHE_ENABLED = 'false'
  process.env.RPC_URL = process.env.RPC_URL || 'https://bsc-dataseed1.binance.org:443'
  process.env.API_VERBOSE = true as unknown as string
  if (process.env.RECORD) {
    nock.recorder.rec()
  }
})

afterAll(() => {
  process.env = oldEnv
  if (process.env.RECORD) {
    nock.recorder.play()
  }

  nock.restore()
  nock.cleanAll()
  nock.enableNetConnect()
})

describe('execute', () => {
  const id = '1'
  let server: http.Server
  let req: SuperTest<Test>

  beforeAll(async () => {
    server = await startServer()
    req = request(`localhost:${(server.address() as AddressInfo).port}`)
    process.env.CACHE_ENABLED = 'false'
  })

  afterAll((done) => {
    server.close(done)
  })

  describe('with from/to', () => {
    const data: AdapterRequest = {
      id,
      data: {
        from: 'USDC',
        to: 'USDT',
      },
    }

    it('should return success', async () => {
      mockEthereumResponseSuccess()

      const response = await req
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })

  describe('with custom params', () => {
    const data: AdapterRequest = {
      id,
      data: {
        from: '0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c',
        fromDecimals: 18,
        to: '0xe9e7cea3dedca5984780bafc599bd69add087d56',
        toDecimals: 18,
        amount: 10,
      },
    }

    it('should return success', async () => {
      mockEthereumResponseSuccess()

      const response = await req
        .post('/')
        .send(data)
        .set('Accept', '*/*')
        .set('Content-Type', 'application/json')
        .expect('Content-Type', /json/)
        .expect(200)
      expect(response.body).toMatchSnapshot()
    })
  })
})
