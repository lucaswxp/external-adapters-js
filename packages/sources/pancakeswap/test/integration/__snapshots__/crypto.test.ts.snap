// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`execute with custom params should return success 1`] = `
Object {
  "data": Object {
    "input": "10000000000000000000",
    "inputDecimals": 18,
    "inputToken": "0xbb4CdB9CBd36B01bD1cBaEBF2De08d9173bc095c",
    "output": "5239325125242389496139",
    "outputDecimals": 18,
    "outputToken": "0xe9e7cea3dedca5984780bafc599bd69add087d56",
    "rate": 523.932512524239,
    "result": 523.932512524239,
  },
  "jobRunID": "1",
  "result": 523.932512524239,
  "statusCode": 200,
}
`;

exports[`execute with from/to should return success 1`] = `
Object {
  "data": Object {
    "input": "1000000000000000000",
    "inputDecimals": 18,
    "inputToken": "0x8ac76a51cc950d9822d68b83fe1ad97b32cd580d",
    "output": "995786777529191071",
    "outputDecimals": 18,
    "outputToken": "0x55d398326f99059ff775485246999027b3197955",
    "rate": 0.9957867775291911,
    "result": 0.9957867775291911,
  },
  "jobRunID": "1",
  "result": 0.9957867775291911,
  "statusCode": 200,
}
`;
